### PR TITLE
Inject runtime error for type mismatch in projection

### PR DESCRIPTION
**Pull request by celinval**
### Description of changes: 

This is to mitigate potentially wrong code generation detected by our type mismatch logic. Eventually we want to enable this assertion and fail at runtime, but this currently affects the compilation of std.

Note: I had to fix the projection error handling for codegen_place. This wasn't working if the failure was in the middle of the projection.

### Resolved issues:

Mitigates #277 

### Call-outs:

The regression will fail until we merge #1073. I was able to verify that once we merge that, this should pass. I won't push until then but I could use some early feedback. 

### Testing:

* How is this change tested?

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs [hunk 1]
```diff
@@ -138,13 +138,13 @@ impl<'tcx> ProjectedPlace<'tcx> {
...
         goto_expr: Expr,
         mir_typ_or_variant: TypeOrVariant<'tcx>,
         fat_ptr_goto_expr: Option<Expr>,
         fat_ptr_mir_typ: Option<Ty<'tcx>>,
         ctx: &mut GotocCtx<'tcx>,
-    ) -> Self {
+    ) -> Result<Self, UnimplementedData> {
         let mir_typ_or_variant = mir_typ_or_variant.monomorphize(ctx);
         let fat_ptr_mir_typ = fat_ptr_mir_typ.map(|t| ctx.monomorphize(t));
         if let Some(fat_ptr) = &fat_ptr_goto_expr {
```
Comment (danielsn) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:150]: Should this also be a return error?

**Review by danielsn — APPROVED** [review]

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs (hunk 1 above)
Comment (celinval) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:150]: Not sure where you are pointing at.

Comment (danielsn) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:150]: We have a mix of `panic` and `Error` in the same function. Was wondering if we should standardize on `Error`

Comment (celinval) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs:150]: I don't think so. Panic is stronger and we should keep doing that if something is really unexpected.

That said, I would be OK compromising and turning those things into debug_assert!() for now so they at least fail in our regression but in the user code, it results in a unsupported_construct.

In the long run, I don't like pushing errors downstream and we should provide stronger guarantees that our compiler works. Everything in this function should eventually trigger an internal compiler error.

**Pull request merged**